docs/blocks-storage/migrate-from-chunks-to-blocks.md

```diff
@@ -0,0 +1,279 @@
+---
+title: "Migrate Cortex cluster from chunks to blocks"
+linkTitle: "Migrate Cortex cluster from chunks to blocks"
+weight: 5
+slug: migrate-cortex-cluster-from-chunks-to-blocks
+---
+
+This article describes how to migrate existing Cortex cluster from chunks storage to blocks storage,
+and highlight possible issues you may encounter in the process.
+
+_This document replaces the [Cortex proposal](https://cortexmetrics.io/docs/proposals/ingesters-migration/),
+which was written before support for migration was in place._
+
+## Introduction
+
+This article **assumes** that:
+
+- Cortex cluster is managed by Kubernetes
+- Cortex is using chunks storage
+- Ingesters are using WAL
+
+_If your ingesters are not using WAL, the documented procedure will still apply, but the presented migration script will not work properly without changes, as it assumes that ingesters are managed via StatefulSet._
+
+The migration procedure is composed by 3 steps:
+
+1. [Preparation](#step-1-preparation)
+1. [Ingesters migration](#step-2-ingesters-migration)
+1. [Cleanup](#step-3-cleanup)
+
+_In case of any issue during or after the migration, this document also outlines a [Rollback](#rollback) strategy._
+
+## Step 1: Preparation
+
+Before starting the migration of ingesters, we need to prepare other services.
+
+### Querier and Ruler
+
+_Everything discussed for querier applies to ruler as well, since it shares querier configuration – options prefix is `querier` even when used by ruler._
+
+Querier and ruler need to be reconfigured as follow:
+
+- `-querier.second-store-engine=blocks`
+- `-querier.query-store-after=0`
+- `-querier.ingester-streaming=false`
+
+#### `-querier.second-store-engine=blocks`
+
+Querier (and ruler) needs to be reconfigured to query both chunks storage and blocks storage at the same time.
+This is achieved by using `-querier.second-store-engine=blocks` option, and providing querier with full blocks configuration, but keeping "primary" store set to `-store.engine=chunks`.
+
+#### `-querier.query-store-after=0`
+
+Querier (and ruler) has an option `-querier.query-store-after` to query store only if query hits data older than some period of time.
+For example, if ingesters keep 12h of data in memory, there is no need to hit the store for queries that only need last 1h of data.
+During the migration, this flag needs to be set to 0, to make queriers always consult the store when handling queries.
+The reason is that after chunks ingesters shutdown, they can no longer return chunks from memory.
```

Comment:
```suggestion
The reason is that, during the ingesters migration, chunks will be flushed to the storage during the shutdown procedure of ingesters running the chunks storage and queriers need to fetch such chunks directly from storage.
```

Reply:
I've accepted all your feedback except this one. I've reworded my text, as I think the real reason is that old ingesters are replaced with new ones, which don't reload chunks from WAL. Please take a look and let me know if it makes sense. Thanks!

Reply:
Your version is waaaaay more clear! 👏

Reply:
Thanks for your help with the PR!